### DISCUSSION TITLE
Compiler complains after fork

### DISCUSSION THREAD

**Discussion by iGavroche** [reach-sh/reach-lang#1133] [Q&A] [answered]
Following a `fork()` which handles one of several possible api calls why does the compiler error out if one of the calls actually `exit()`s before setting some of the variables required post-fork in the cases where we would not `exit()` ?
example:
```
fork.api(caller1.leave, (<no arguments>)... <eventually uses exit()>).api(caller2.buyGizmo, (gizmoId));

assert(balance(gizmoId) == 1, 'what happened to gizmo?');
```

The compiler would error out on the `assert `because `gizmoId `may not exist, but `caller1.leave` does exit() during API Consensus step and `gizmoId` is never part of the method's arguments nor is it instantiated.

**Comment by jeapostrophe** [comment, marked as answer]
Reach has normal lexical scope in this case. If you want to use the identifier binding, you need to put the code in the API case. If you want to use the value, then I recommend that you `return`.

**Reply by jeapostrophe** [reply]
You may also be interested in using `throw` for the exceptional case.

**Reply by iGavroche** [reply]
thank you very much, @jeapostrophe. I tried returning from a fork as such:

```
const myResult = fork()
  .api(myAPI.callItQuits,
    () => { assume(1); },
    (returnFunc) => {
      returnFunc(true);

      commit();
      exit();
    }
  )
  .api(myOtherAPI.callMethod, 
    (arg1, arg2) => { assume(1); }, 
    (arg1, arg2, retFunc) => { 
      // returns to API Caller
      retFunc([arg1, arg2]); 

      // Gets https://docs.reach.sh/rsh/errors/#RE0093
      return [arg1, arg2]; 
    })
```

I'm not sure I can return from a fork as we can from a parallelReduce(). Do you have any pointer?

**Reply by jeapostrophe** [reply]
You can only return from functions... so make a function!

```
const myResult = (() => fork()
.api(myAPI.callItQuits,
    () => { assume(1); },
    (returnFunc) => {
      returnFunc(true);

      commit();
      exit();
    }
  )
  .api(myOtherAPI.callMethod, 
    (arg1, arg2) => { assume(1); }, 
    (arg1, arg2, retFunc) => { 
      // returns to API Caller
      retFunc([arg1, arg2]); 

      // Gets https://docs.reach.sh/rsh/errors/#RE0093
      return [arg1, arg2]; 
    }))();
```
